File: example/calculator/native/Sources/Calculator/Calculator.swift
import Swiftler
import Foundation

#nifLibrary(name: "calculator", functions: [
    add(_:_:), 
    subtract(_:_:), 
    multiply(_:_:), 
    divide(_:_:),
    power(_:_:),
    sqrt(_:),
    factorial(_:),
    is_prime(_:),
    gcd(_:_:),
    fibonacci(_:),
    circle_area(_:),
    greet(_:)
])

// Basic arithmetic operations

@nif func add(_ a: Int, _ b: Int) -> Int {
    a + b
}

@nif func subtract(_ a: Int, _ b: Int) -> Int {
    a - b
}

@nif func multiply(_ a: Int, _ b: Int) -> Int {
    a * b
}

@nif func divide(_ a: Int, _ b: Int) -> Int {
    guard b != 0 else { return 0 }
    return a / b
}

// Advanced mathematical operations

@nif func power(_ base: Int, _ exponent: Int) -> Int {
    Int(pow(Double(base), Double(exponent)))
}

@nif func sqrt(_ n: Double) -> Double {
    Foundation.sqrt(n)
}

@nif func factorial(_ n: Int) -> Int {
    guard n >= 0 else { return 0 }
    guard n > 1 else { return 1 }
    return (2...n).reduce(1, *)
}

@nif func is_prime(_ n: Int) -> Bool {
    guard n > 1 else { return false }
    guard n != 2 else { return true }
    guard n % 2 != 0 else { return false }
    
    let sqrtN = Int(Foundation.sqrt(Double(n)))
    for i in stride(from: 3, through: sqrtN, by: 2) {
        if n % i == 0 {
            return false
        }
    }
    return true
}

@nif func gcd(_ a: Int, _ b: Int) -> Int {
    let absA = abs(a)
    let absB = abs(b)
    var x = absA
    var y = absB
    
    while y != 0 {
        let temp = y
        y = x % y
        x = temp
    }
    
    return x
}

@nif func fibonacci(_ n: Int) -> Int {
    guard n >= 0 else { return 0 }
    guard n > 1 else { return n }
    
    var a = 0
    var b = 1
    
    for _ in 2...n {
        let temp = a + b
        a = b
        b = temp
    }
    
    return b
}

@nif func circle_area(_ radius: Double) -> Double {
    Double.pi * radius * radius
}

// String operations

@nif func greet(_ name: String) -> String {
    "Hello, \(name)! Welcome to Swiftler Calculator."
}

File: example/calculator/lib/calculator.ex
defmodule Calculator do
  @moduledoc """
  A calculator library powered by Swift NIFs via Swiftler.
  
  This module demonstrates how to integrate Swift code with Elixir using Swiftler.
  It provides various mathematical operations implemented in Swift.
  """

  use Swiftler, otp_app: :calculator

  # Basic arithmetic operations
  swift_function add(a: :int, b: :int) :: :int
  swift_function subtract(a: :int, b: :int) :: :int
  swift_function multiply(a: :int, b: :int) :: :int
  swift_function divide(a: :int, b: :int) :: :int

  # Advanced mathematical operations
  swift_function power(base: :int, exponent: :int) :: :int
  swift_function sqrt(n: :double) :: :double
  swift_function factorial(n: :int) :: :int
  swift_function is_prime(number: :int) :: :bool
  swift_function gcd(a: :int, b: :int) :: :int
  swift_function fibonacci(n: :int) :: :int
  swift_function circle_area(radius: :double) :: :double

  # String operations
  swift_function greet(name: :string) :: :string

  @doc """
  Calculates the least common multiple of two integers.
  
  ## Examples
  
      iex> Calculator.lcm(12, 18)
      36
      
      iex> Calculator.lcm(7, 5)
      35
  """
  def lcm(a, b) when is_integer(a) and is_integer(b) do
    div(abs(a * b), gcd(a, b))
  end

  @doc """
  Generates a list of prime numbers up to n.
  
  ## Examples
  
      iex> Calculator.primes_up_to(20)
      [2, 3, 5, 7, 11, 13, 17, 19]
  """
  def primes_up_to(n) when is_integer(n) and n >= 2 do
    2..n
    |> Enum.filter(&is_prime/1)
  end

  @doc """
  Generates the first n Fibonacci numbers.
  
  ## Examples
  
      iex> Calculator.fibonacci_sequence(10)
      [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
  """
  def fibonacci_sequence(n) when is_integer(n) and n >= 1 do
    0..(n - 1)
    |> Enum.map(&fibonacci/1)
  end

  @doc """
  Calculates the volume of a sphere given its radius.
  
  ## Examples
  
      iex> Calculator.sphere_volume(3.0)
      113.09733552923255
  """
  def sphere_volume(radius) when is_number(radius) do
    # Convert to integer for power function, then back to float
    radius_cubed = power(round(radius), 3)
    4.0 / 3.0 * :math.pi() * radius_cubed
  end

  @doc """
  Evaluates a simple mathematical expression.
  
  ## Examples
  
      iex> Calculator.evaluate("2 + 3 * 4")
      14
  """
  def evaluate(expression) when is_binary(expression) do
    # This is a simple example - in production, use a proper parser
    case String.split(expression, ~r/\s+/) do
      [a, "+", b] -> add(String.to_integer(a), String.to_integer(b))
      [a, "-", b] -> subtract(String.to_integer(a), String.to_integer(b))
      [a, "*", b] -> multiply(String.to_integer(a), String.to_integer(b))
      [a, "/", b] -> divide(String.to_integer(a), String.to_integer(b))
      _ -> {:error, "Unsupported expression"}
    end
  end
end